man/cerebro.1
.\"#############################################################################
.\"$Id: cerebro.1,v 1.3 2008-03-28 17:06:47 chu11 Exp $
.\"#############################################################################
.\"  Copyright (C) 2007-2008 Lawrence Livermore National Security, LLC.
.\"  Copyright (C) 2005-2007 The Regents of the University of California.
.\"  Produced at Lawrence Livermore National Laboratory (cf, DISCLAIMER).
.\"  Written by Albert Chu <chu11@llnl.gov>.
.\"  UCRL-CODE-155989 All rights reserved.
.\"
.\"  This file is part of Cerebro, a collection of cluster monitoring tools
.\"  and libraries.  For details, see <http://www.llnl.gov/linux/cerebro/>.
.\"
.\"  Cerebro is free software; you can redistribute it and/or modify it under
.\"  the terms of the GNU General Public License as published by the Free
.\"  Software Foundation; either version 2 of the License, or (at your option)
.\"  any later version.
.\"
.\"  Cerebro is distributed in the hope that it will be useful, but WITHOUT ANY
.\"  WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
.\"  FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
.\"  details.
.\"
.\"  You should have received a copy of the GNU General Public License along
.\"  with Cerebro.  If not, see <http://www.gnu.org/licenses/>.
.\"#############################################################################
.TH cerebro 1 "Oct 2007" "cerebro" cerebro
.SH "NAME"
cerebro \- cluster monitoring tools and libraries
.SH "DESCRIPTION"
.B Cerebro
is a collection of monitoring tools and libraries.  It is made up of a
daemon, loadable modules, a library, and several tools.

The daemon, 
.B cerebrod(8),
is the core of
.B Cerebro.
Typically executed on every node of a cluster, it is responsible for
sending, receiving, and monitoring various metric data.  

The data that it sends, receives, and monitors is dependent on the 
data configured to be sent.  Data can be sent via modules
developed for 
.B cerebrod,
(See 
.B cerebro_modules(3)
for more information on these modules.), data fed into the daemon
using 
.B libcerebro(3),
or data fed into the daemon via 
.B cerebro-admin.

.B Libcerebro
and 
.B cerebro-admin
offer the user the ability to monitor data by programming them into tools
and/or scripts.  Modules developed for 
.B cerebrod,
will be automatically loaded by the daemon and those specific metrics
will automatically be monitored.

Before
.B cerebro
can be used as a whole it must be configured.  The configuration file
options are specified in 
.B cerebro.conf(5).
Generally speaking, all cluster nodes should be configured to 'speak' so
they can send out metric information.  A few cluster nodes, typically
management nodes, should be configured to 'listen' and gather this metric information.
The 'metric_controller' should be enabled if data will be fed into 
.B cerebrod
via
.B libcerebro
and/or 
.B cerebro-admin.
The 'metric_server' should be enabled if data will be read
from
.B cerebrod
via
.B libcerebro
and/or
.B cerebro-stat.
A configuration module is also available for use with
.B cerebrod.
If your orgnization already has a database or configuration system for
clusters, the configuration module may allow your orgnization to 
configure 
.B cerebrod
off that database instead.

Three other module types are available for use in 
.B cerebrod.
 'clusterlist' modules define what nodes exist within a cluster.
 'monitor' modules allow specific metrics to be passed to the 
module for handling, perhaps for logging purposes.
 'event' modules allow for the creation of events when certain
metric information is received.
See 
.B cerebro_modules(3)
for more details.

An event notification API is provided with 
.B libcerebro
for programming against these events.

.SH "ORIGIN"
Developed by Albert Chu <chu11@llnl.gov> on LLNL's Linux clusters.
.SH "SEE ALSO"
libcerebro(3), cerebro_modules(3), cerebrod(8)


NEWS
This file describes changes in recent versions of cerebro. It
primarily documents those changes that are of interest to users and
admins.

Changes in cerebro-1.5-1
===========================
 -- Fix mem-corruption bug.

Changes in cerebro-1.4-1
===========================
 -- Various bug fixes.

Changes in cerebro-1.3-1
===========================
 -- Redo configuration file format for message listening and speaking.
 -- Support message forwarding.
 -- Support module versioning.

Changes in cerebro-1.2-1
===========================
 -- Documentation updates

Changes in cerebro-1.1-1
===========================
 -- Allow cerebro to work on multiple subnets.
 -- Added load average, memory, and network traffic metric modules.
 -- Support event based monitoring.
 -- Various bug fixes.

Changes in cerebro-1.0-4
===========================
 -- Added graceful shutdown metric module.
 -- Allow monitor modules to support multiple metrics.
 -- Various bug fixes.

Changes in cerebro-1.0-1
===========================
 -- Initial Release

$Id: NEWS,v 1.11 2008-04-01 22:01:46 chu11 Exp $


src/libs/cerebro/cerebro.c
/*****************************************************************************\
 *  $Id: cerebro.c,v 1.18 2008-03-28 17:06:48 chu11 Exp $
 *****************************************************************************
 *  Copyright (C) 2007-2008 Lawrence Livermore National Security, LLC.
 *  Copyright (C) 2005-2007 The Regents of the University of California.
 *  Produced at Lawrence Livermore National Laboratory (cf, DISCLAIMER).
 *  Written by Albert Chu <chu11@llnl.gov>.
 *  UCRL-CODE-155989 All rights reserved.
 *
 *  This file is part of Cerebro, a collection of cluster monitoring
 *  tools and libraries.  For details, see
 *  <http://www.llnl.gov/linux/cerebro/>.
 *
 *  Cerebro is free software; you can redistribute it and/or modify it under
 *  the terms of the GNU General Public License as published by the Free
 *  Software Foundation; either version 2 of the License, or (at your option)
 *  any later version.
 *
 *  Cerebro is distributed in the hope that it will be useful, but WITHOUT ANY
 *  WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 *  FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
 *  details.
 *
 *  You should have received a copy of the GNU General Public License along
 *  with Cerebro.  If not, see <http://www.gnu.org/licenses/>.
\*****************************************************************************/

#if HAVE_CONFIG_H
#include "config.h"
#endif /* HAVE_CONFIG_H */

#include <stdio.h>
#include <stdlib.h>
#if STDC_HEADERS
#include <string.h>
#endif /* STDC_HEADERS */

#include "cerebro.h"
#include "cerebro_api.h"
#include "cerebro_config_util.h"
#include "cerebro_util.h"
#include "cerebro/cerebro_config.h"

#include "debug.h"

static char *cerebro_error_messages[] =
  {
    "success",                                   /* 0 */
    "null cerebro_t handle",                     /* 1 */
    "null cerebro_namelist_t namelist",          /* 2 */
    "null cerebro_namelist_iterator_t iterator", /* 3 */
    "null cerebro_nodelist_t nodelist",          /* 4 */
    "null cerebro_nodelist_iterator_t iterator", /* 5 */
    "invalid magic number found",                /* 6 */
    "invalid parameters",                        /* 7 */
    "invalid hostname",                          /* 8 */
    "connection error",                          /* 9 */
    "connection timeout",                        /* 10 */
    "protocol error",                            /* 11 */
    "protocol timeout",                          /* 12 */
    "version incompatible",                      /* 13 */
    "buffer overflow",                           /* 14 */
    "node not found",                            /* 15 */
    "metric name invalid",                       /* 16 */
    "event name invalid",                        /* 17 */
    "metric max reached",                        /* 18 */
    "end of list reached",                       /* 19 */
    "event not received",                        /* 20 */
    "config file error",                         /* 21 */
    "config module error",                       /* 22 */
    "config input error",                        /* 23 */
    "out of memory",                             /* 24 */
    "internal error",                            /* 25 */
    "errnum out of range",                       /* 26 */
  };

cerebro_t
cerebro_handle_create(void)
{
  cerebro_t handle = NULL;

  if (!(handle = (cerebro_t)malloc(sizeof(struct cerebro))))
    goto cleanup;
                 
  memset(handle, '\0', sizeof(struct cerebro));
  handle->magic = CEREBRO_MAGIC_NUMBER;
  handle->errnum = CEREBRO_ERR_SUCCESS;
  memset(&(handle->hostname), '\0', CEREBRO_MAX_HOSTNAME_LEN+1);
  handle->port = 0;
  handle->timeout_len = 0;
  handle->flags = 0;
  handle->loaded_state = 0;
  memset(&(handle->config_data), '\0', sizeof(struct cerebro_config));
  
  if (!(handle->namelists = list_create((ListDelF)cerebro_namelist_destroy)))
    goto cleanup;

  if (!(handle->nodelists = list_create((ListDelF)cerebro_nodelist_destroy)))
    goto cleanup;

  if (!(handle->event_fds = list_create((ListDelF)free)))
    goto cleanup;

  return handle;

 cleanup:
  if (handle)
    {
      if (handle->namelists)
        list_destroy(handle->namelists);
      if (handle->nodelists)
        list_destroy(handle->nodelists);
      if (handle->event_fds)
        list_destroy(handle->event_fds);
      free(handle);
    }
  return NULL;
}

int
cerebro_handle_destroy(cerebro_t handle)
{
  if (_cerebro_handle_check(handle) < 0)
    return -1;

  if (handle->loaded_state & CEREBRO_CONFIG_LOADED)
    {
      if (_cerebro_unload_config(handle) < 0)
	return -1;

      if (handle->loaded_state & CEREBRO_CONFIG_LOADED)
        {
          CEREBRO_DBG(("loaded_state invalid"));
          handle->errnum = CEREBRO_ERR_INTERNAL;
          return -1;
        }
    }

  list_destroy(handle->namelists);
  handle->namelists = NULL;
  list_destroy(handle->nodelists);
  handle->nodelists = NULL;
  list_destroy(handle->event_fds);
  handle->event_fds = NULL;
  
  handle->errnum = CEREBRO_ERR_SUCCESS;
  handle->magic = ~CEREBRO_MAGIC_NUMBER;
  free(handle);
  return 0;
}

int
cerebro_errnum(cerebro_t handle)
{
  if (!handle)
    return CEREBRO_ERR_NULLHANDLE;
  else if (handle->magic != CEREBRO_MAGIC_NUMBER)
    return CEREBRO_ERR_MAGIC_NUMBER;
  else
    return handle->errnum;
}

int 
cerebro_namelist_errnum(cerebro_namelist_t namelist)
{
  if (!namelist)
    return CEREBRO_ERR_NULLNAMELIST;
  else if (namelist->magic != CEREBRO_NAMELIST_MAGIC_NUMBER)
    return CEREBRO_ERR_MAGIC_NUMBER;
  else
    return namelist->errnum;
}

int 
cerebro_namelist_iterator_errnum(cerebro_namelist_iterator_t namelistItr)
{
  if (!namelistItr)
    return CEREBRO_ERR_NULLNAMELIST_ITERATOR;
  else if (namelistItr->magic != CEREBRO_NAMELIST_ITERATOR_MAGIC_NUMBER)
    return CEREBRO_ERR_MAGIC_NUMBER;
  else
    return namelistItr->errnum;
}

int 
cerebro_nodelist_errnum(cerebro_nodelist_t nodelist)
{
  if (!nodelist)
    return CEREBRO_ERR_NULLNODELIST;
  else if (nodelist->magic != CEREBRO_NODELIST_MAGIC_NUMBER)
    return CEREBRO_ERR_MAGIC_NUMBER;
  else
    return nodelist->errnum;
}

int 
cerebro_nodelist_iterator_errnum(cerebro_nodelist_iterator_t nodelistItr)
{
  if (!nodelistItr)
    return CEREBRO_ERR_NULLNAMELIST_ITERATOR;
  else if (nodelistItr->magic != CEREBRO_NODELIST_ITERATOR_MAGIC_NUMBER)
    return CEREBRO_ERR_MAGIC_NUMBER;
  else
    return nodelistItr->errnum;
}

char *
cerebro_strerror(int errnum)
{
  if (errnum >= CEREBRO_ERR_SUCCESS && errnum <= CEREBRO_ERR_ERRNUMRANGE)
    return cerebro_error_messages[errnum];
  else
    return cerebro_error_messages[CEREBRO_ERR_ERRNUMRANGE];
}

char *
cerebro_get_hostname(cerebro_t handle)
{
  if (_cerebro_handle_check(handle) < 0)
    return NULL;

  return handle->hostname;
}

int 
cerebro_set_hostname(cerebro_t handle, const char *hostname)
{
  if (_cerebro_handle_check(handle) < 0)
    return -1;

  if (hostname && strlen(hostname) > CEREBRO_MAX_HOSTNAME_LEN)
    {
      handle->errnum = CEREBRO_ERR_OVERFLOW;
      return -1;
    }

  if (hostname)
    strcpy(handle->hostname, hostname);
  else
    memset(handle->hostname, '\0', CEREBRO_MAX_HOSTNAME_LEN+1);

  return 0;
}

int 
cerebro_get_port(cerebro_t handle)
{
  if (_cerebro_handle_check(handle) < 0)
    return -1;

  return handle->port;
}

int 
cerebro_set_port(cerebro_t handle, unsigned int port)
{
  if (_cerebro_handle_check(handle) < 0)
    return -1;

  handle->port = port;
  return 0;
}

int 
cerebro_get_timeout_len(cerebro_t handle)
{
  if (_cerebro_handle_check(handle) < 0)
    return -1;

  return handle->timeout_len;
}

int 
cerebro_set_timeout_len(cerebro_t handle, unsigned int timeout_len)
{
  if (_cerebro_handle_check(handle) < 0)
    return -1;

  handle->timeout_len = timeout_len;
  return 0;
}

int 
cerebro_get_flags(cerebro_t handle)
{
  if (_cerebro_handle_check(handle) < 0)
    return -1;

  return handle->flags;
}

int 
cerebro_set_flags(cerebro_t handle, unsigned int flags)
{
  if (_cerebro_handle_check(handle) < 0)
    return -1;

  if (flags & ~CEREBRO_METRIC_FLAGS_MASK)
    {
      handle->errnum = CEREBRO_ERR_PARAMETERS;
      return -1;
    }

  handle->flags = flags;
  return 0;
}
